File: leetcode/easy/ClimbingStairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        """
        Calculates the number of distinct ways to climb to the top of n stairs.
        You can either climb 1 or 2 steps at a time.

        This problem is a classic example of dynamic programming and is equivalent
        to calculating the (n+1)-th Fibonacci number.

        Let ways(i) be the number of ways to reach step i.
        ways(i) = ways(i-1) + ways(i-2)

        Base cases:
        ways(1) = 1 (one step)
        ways(2) = 2 (one step + one step, or two steps)

        This solution uses a bottom-up approach with O(1) space.
        """
        # We only need to store the ways to reach the previous two steps.
        # two_steps_back: ways to reach step i-2. Initialize for step 0 (1 way: do nothing).
        # one_step_back: ways to reach step i-1. Initialize for step 1 (1 way).
        two_steps_back, one_step_back = 1, 1

        # Iterate from step 2 up to n. This loop runs n-1 times.
        for _ in range(n - 1):
            two_steps_back, one_step_back = one_step_back, two_steps_back + one_step_back

        # After the loop, one_step_back holds the number of ways for step `n`.
        return one_step_back

File: leetcode/easy/ClimbingStairs.java
class Solution {
    /**
     * Calculates the number of distinct ways to climb to the top of a staircase.
     *
     * This problem is a classic example of dynamic programming and is equivalent
     * to generating the Fibonacci sequence.
     *
     * Let `dp[i]` be the number of ways to reach the i-th step.
     * To reach step `i`, you can either take one step from `i-1` or two steps from `i-2`.
     * Therefore, the recurrence relation is: `dp[i] = dp[i-1] + dp[i-2]`.
     *
     * This solution uses a space-optimized DP approach. Instead of storing all
     * previous results in an array, we only need to keep track of the last two
     * values (the ways to reach the previous step and the step before that).
     *
     * Time Complexity: O(n), as we iterate from 3 to n once.
     * Space Complexity: O(1), as we only use a few variables to store state.
     *
     * @param n The total number of steps in the staircase.
     * @return The total number of distinct ways to climb to the top.
     */
    public int climbStairs(int n) {
        // Base cases:
        // If n=0, there is 1 way (do nothing).
        // If n=1, there is 1 way (1 step).
        // If n=2, there are 2 ways (1+1 or 2 steps).
        if (n <= 2) {
            return n;
        }

        // `twoStepsBack` stores the number of ways to reach step i-2.
        // Initialize with the ways to reach step 1.
        int twoStepsBack = 1;
        
        // `oneStepBack` stores the number of ways to reach step i-1.
        // Initialize with the ways to reach step 2.
        int oneStepBack = 2;

        // Iterate from the 3rd step up to n.
        for (int i = 3; i <= n; i++) {
            // The number of ways to reach the current step `i` is the sum of the ways
            // to reach the previous two steps.
            int currentWays = oneStepBack + twoStepsBack;
            
            // Update the pointers for the next iteration.
            twoStepsBack = oneStepBack;
            oneStepBack = currentWays;
        }

        // `oneStepBack` now holds the number of ways to reach the n-th step.
        return oneStepBack;
    }
}
